imapserver/session.go

```diff
@@ -124,3 +124,11 @@ type SessionAppendLimit interface {
 	// this server in an APPEND command.
 	AppendLimit() uint32
 }
+
+// SessionNotify is an IMAP session which supports NOTIFY.
+type SessionNotify interface {
+	Session
+
+	// Authenticated state
+	Notify(w *UpdateWriter, options *imap.NotifyOptions) error
```

Comment:
I wonder about the implications of passing an `UpdateWriter` to the backend here. Up until now backends would only use `UpdateWriter` synchronously in handlers. Here the backend would need to use it after `Notify` returns. We have `conn.encMutex` to avoid stepping on our own toes when sending an update, but that still leaves considerations such as sequence number issues with EXPUNGE notifications while another command is running, see [section 5.5](https://www.rfc-editor.org/rfc/rfc9051.html#section-5.5).

Maybe there is a way to expose a safer interface to backends. Maybe we should just leave it up to the backend and document the issue.

Reply:
FWIW, I only wrote some of the server portions to test the client, but eventually moved on to just testing with dovecot.